### ISSUE TITLE
Proposal: BERTopic in Weaviate for Cluster Analysis

### ISSUE THREAD

**Issue by CShorten** [semi-technologies/weaviate#2304]
### What
Large text datasets tend to be composed of many different topics. For example, Wikipedia contains text about animals, sports, medicine, and many others. We can use clusters in vector space to extract these high-level topics automatically from unstructured text data.
### Why
BERTopic will enable Weaviate users to understand the semantic clusters in their text datasets by concatenating each cluster into “long” documents. We then apply keyword importance scoring to the long documents, such as TF-IDF, and return a list of keywords describing each semantic cluster. 

Clusters are calculated by first compressing high-dimensional vectors with UMAP or t-SNE (such as 384-d to 15-d). The clusters are then identified from the lower-dimensional vectors with HDBSCAN. As mentioned before, these clusters are then concatenated into long documents and the TF-IDF most important keyword lists are returned to the user.

**Recommended articles:**
<a href = "https://medium.com/towards-data-science/topic-modeling-with-bert-779f7db187e6">Topic Modeling with BERT:</a> This article contains a step-by-step implementation of BERTopic in python from scratch.
<a href = "https://towardsdatascience.com/interactive-topic-modeling-with-bertopic-1ea55e7d73d8">Interactive Topic Modeling with BERTopic:</a> This article describes some of the features accessible in the BERTopic python library.

**The “GroupBy” analog in Vector Space**

Cluster Analysis is extremely useful in Machine Learning for detecting distribution shifts, constructing generalization tests, and generally understanding massive datasets that are incomprehensible to comb through row-by-row. We can extend BERTopic from text only to using embeddings derived from images and their associated text descriptions to form topics, and many other ideas as multimodal text-X applications continue to evolve.

I think this category of Cluster Analysis is the Vector Space equivalent of GroupBy analysis with symbolic attributes, which has been super useful in SQL-style systems and so on. However, instead of “GroupBy” with symbol matching, we use clusters in the vector space.

### How

BERTopic operates in 4 steps:

1. Vectorize the dataset (Weaviate of course already has these vectors stored)

2. Compress high-dimensional vectors into lower-dimensional vectors. Lower-dimensional compression in BERTopic is typically done with UMAP, but I found a nice golang library for t-SNE that Etienne has contributed to - https://github.com/danaugrs/go-tsne. The library utilizes the gonum/mat library. I think it makes more sense to use this t-SNE implementation, and then test t-SNE vs. UMAP in python later on and see if it is worth the extra effort.

3. The clustering algorithm used in Step 3 (HDBSCAN) won’t work with 384-d vectors. We could skip this by clustering with k-Means, which may work for topic clustering -- but not outlier detection as a side effect, I don’t think it makes sense to use k-Means over t-SNE / UMAP → HDBSCAN.

Cluster the lower dimensional embeddings with HDBSCAN – I  found this golang library for it: https://pkg.go.dev/github.com/humilityai/hdbscan#section-readme

```golang
// go get -u github.com/humilityai/hdbscan

import(
  "github.com/humilityai/hdbscan"
)

func main() {
  data := [][]float64{
                  []float64{1,2,3},
                  []float64{3,2,1},
               }
  minimumClusterSize := len(data)
  minimumSpanningTrue := true

  // create
  clustering, err := hdbscan.NewClustering(data, minimumClusterSize)
  if err != nil {
    panic(err)
  }

// options
clustering = clustering.Verbose().OutlierDetection() // we may want to use this as well for Outlier Detection in Weaviate

// run
clustering.Run(hdbscan.EuclideanDistance, hdbscan.VarianceScore, minimumSpanningTree)

// We can use the Assign() method afterwards on the total dataset
```

We then need to call Assign() afterwards to label each document with the cluster it belongs to, info on the Assign method starts on line 51 here: https://github.com/humilityai/hdbscan/blob/master/sampling.go.

4. We then concatenate all the documents in a unique cluster to form a CLUSTER DOCUMENT! This is done using the labels derived from the Assign() method from the hdbscan library output. We then apply tf-idf to the set of cluster documents. Here is another golang library that implements tf-idf - https://github.com/wilcosheh/tfidf/blob/master/tfidf.go. Once TF-IDF is completed, we sort the keywords in descending rank order and report the number of keywords / topics specified in the UX design.

### BERTopic - Weaviate UX Design

The output of BERTopic is a list of Topics. Each Topic contains a: (1) list of keywords (2) list of scores associated with each keyword and (k) properties we may want to add. For example, the BERTopic python library stores (3) number of documents in the Topic cluster, (4) representative docs at the centroid of the cluster. We may also want to add (5) wordCount statistics as used in previous Aggregate queries. An illustration of this is shown below with class name in bold and properties in normal font.

Since the output of BERTopic is meta information, rather than something each class instance should “carry” with it. I think it makes the most sense to form New Classes that store this information, i.e:

<img width="528" alt="BERTopic-new-class" src="https://user-images.githubusercontent.com/25864937/196186720-88f9dadd-d970-4cbe-82f1-a714caa12b16.png">

I think this is a nice way to flex the muscles of Weaviate’s relational design. 

When we send the run reust to the BERTopic in a Weaviate Class constructor by telling it (1) which property or properties to use in TF-IDF (must be string / text type) and (2) how many centroidDocs to store.

centroidDocs are the vectors in the center of the cluster, they are calculated by applying k-means to each of the clusters separately. Here is a golang library that implements k-means clustering - https://pkg.go.dev/github.com/biogo/cluster/kmeans. After running this, we can access the centroids with:

```golang
func (km *Kmeans) Centers() []cluster.[Center](https://pkg.go.dev/github.com/biogo/cluster@v0.0.0-20150317015930-38d63f016191/cluster#Center)
```

Clustering is done using the 1 vector to 1 data object design in Weaviate. 

BERTopic configuration is either set up via ENABLE_MODULES or adding “BERTopic” to the “moduleConfig” when constructing a schema with one of weaviate’s clients. The arguments for BERTopic are passed in the run call rather than in the moduleConfig.

```python
“moduleConfig”: {
	“BERTopic”, #no arguments are passed to BERTopic in moduleConfig
	…
}
```

**When does BERTopic run?**

The BERTopic module will run upon request from a weaviate client function such as the following python method. This is needed because we don’t want to recompute BERTopic every time a new object is created and we also want to allow the user to customize the topics, topK, and centroidDocs arguments.

```python
'''
Input arguments:
  -- class_name   (str): The Weaviate Class to apply BERTopic to
  -- path         (str): The path to the text used in forming the long documents for keyword scoring
  -- topics       (int): The number of topics to write to the new object formed by BERTopic 
  -- topK         (int): The number of top keywords to write to each topic in the new object formed by BERTopic
  -- centroidDocs (int): The number of centroidDocs (centers derived from k-means in each cluster) to write to the new object
'''
client.bertopic.run("Article", "content", 5, 10, 3)
```

BERTopic will use the class and path to name the new object that store topics information – i.e. ArticleContentTopic from the example above. As Weavaite receives the arguments to a BERTopic call it will first check if this object already exists and if so, delete it. The uuid can be a random number sampled from a massive range to avoid collision with another object.

Thinking the documentation of BERTopic client should  look similar to: https://weaviate-python-client.readthedocs.io/en/stable/weaviate.classification.html. 

Note that calling BERTopic like this is different from accessing the module in a Weaviate GET query – this is shown later in the proposal in Query Example #3.

### Query Example 1

We can then access topic information for the entire Article Class broadly through:

```graphql
{
	Get {
		ArticleContentTopic {
			keywords
			scores
			centroidDocs
		}
	}
}
```

Which will return:

```python
{
	"ArticleContentTopic": {
		"keywords": [
			["medicine", "healthcare", …],
			["basketball", "baseball", …],
			["golden retriever", "tiger", …],
		],
		"scores": [
			[1.2344, 1.043, …],
			[8.1, 4.2, …],
			[1.0002, 0.65, …],
		],
		"centroidDocs": [
			["Medicine costs have been skyrocketing", "Hospitals awaiting shipment of medicine in Boston",…],
                        ["There are 32 NBA basketball teams", "Baseball is the oldest sport in America",…],
                        ["Golden retrievers are the most common dog breed in the world", "800 pound Tiger arrives at Bronx Zoo", ...],	
		]
	}
}
```

### Query Example 2 

We can also look at topic information for an individual Document through the cross-reference query syntax, i.e:

```graphql
{
	Get {
		Article {
			ArticleContentTopicID
			... on ArticleContentTopic {
				keywords
				scores
				centroidDocs
			}
		}
	}
}
```

Which will return:

```python
{
	"Article": {
		"hasContentTopicID": [
			{
				"keywords": ["medicine", "healthcare", …],
				"scores":  [1.2344, 1.043, …],
				"centroidDocs": ["Medicine costs have been skyrocketing", 
						"Hospitals awaiting shipment of medicine in Boston", …],
			}
		]
	}
}
```

### Query Example 3

In addition to applying BERTopic on the entire Class, we may also want to subset the Class with a query or filter and then apply topic modeling from there. When the data is zoomed in like this, HDBSCAN will return more fine-grained clusters compared to looking at the entire dataset:

<img width="615" alt="BERTopic-subset-query" src="https://user-images.githubusercontent.com/25864937/196187506-6e1aeb81-1bd3-4992-90b4-dc4de61fccea.png">

In this case, we envision accessing the topic lists through the _additional design, such as:

```graphql
{
	Get {
		Article (
			nearText: {
				concepts: ["Medications for Insomnia"]
			}
			BERTopic: {
				path: ["title"],
				topics: 5,
				topK: 10,
				centroidDocs: 3
			}
		){
		year
		title
		_additional {
			keyword_lists,
			score_lists,
			centroidDocs
			}
		}
	}
				
}
```

Which will return:

```python
{
	"Article": {
		"year": 2020,
		"title": "COVID-19 outbreak in Washington",
		"_additional": {
			"keyword_lists": [
				["pandemic", "spread", "epidemiology", ...],
				["convalescent", "plasma", "therapy", ...],
			],
			"score_lists": [
				[0.8, 0.6, 0.4, ..],
				[1.4, 1.3, 1.2, ..],
			],
			"centroidDocs": [
				["The pandemic has spread to Washington, USA", ...],
				["Convalescent plasma therapy is undergoing clinical trials", ...],
			]
		}
	}
}
```

So the entire Topic set is then appended to each returned item. I wasn't able to think of another way of doing this, of course open to suggestions.

### Additional Considerations

1. BERTopic in a python container to run with RAPIDS GPU implementation

Tl;dr - RAPIDS is a python library by NVIDIA that is pretty much pandas dataframes for CUDA kernels, the cudf objects offer a pretty much identical functionality to the pandas DataFrame. RAPIDS also has cuml 

Here is an old video I made on RAPIDS - “600x Speedup Example t-SNE on RAPIDS”: https://www.youtube.com/watch?v=_4OehmMYr44. Pretty cool stuff, super fast -- of course users need a GPU for this (preferably more than one as well).

Probably best to get an MVP out there in golang as proposed, see how popular it is and then consider adding this. It could just be like a “Do you have a GPU?” -- Yes -- try BERTopic-RAPIDS, which would be a custom module paired with a python container.

### Acknowledgements

Thank you to Laura Ham and Bob van Luijt for help in developing this proposal!

**Comment by beckernick** [comment]
Hi! I came across this issue due to the cuML/RAPIDS mention.

I just wanted to mention that RAPIDS [now provides pip packages](https://rapids.ai/pip.html) in case it streamlines your potential container environment. Happy to try to answer any other questions about RAPIDS that might come up.

**Comment by CShorten** [comment]
@donomii, I would really appreciate your thoughts on this if you have a chance.

**Comment by ssuygun** [comment]
Hierarchical Clustering - We would love to be able to have a module that is capable of reducing the entire or subset of the dataset to then apply clustering algo's over the reduced dimensions. We have all this searchable data but would love a way to suggest a taxonomy of our data in an unsupervised manner. Can we work around here?

**Comment by CShorten** [comment]
Note, please see discussion from Juraj Bezdek and Leland McInnes on distributed UMAP for BERTopic in Weaviate - https://twitter.com/leland_mcinnes/status/1594090430400196608.

**Comment by CShorten** [comment]
Requested on Twitter - https://twitter.com/danlovesproofs/status/1620191061804187648.
